SBOM generator

atomicturtle opened the pull request:
# Pull Request: Add SBOM Generator Plugin

## Summary
This PR adds a new plugin, `sbom_generator`, which automatically generates a Software Bill of Materials (SBOM) for packages built with Mock.

## How it Works
The plugin runs as a `postbuild` hook within the Mock lifecycle. It performs the following actions:
- **Introspection**: Queries the build chroot for active RPM macros to identify build environment configuration.
- **Metadata Extraction**: Extracts package headers, file lists, SHA-256 hashes, and GPG signatures from binary RPMs and SRPMs.
- **Dependency Resolution**: Maps runtime dependencies and optionally the build toolchain installed in the chroot.
- **Standardization**: Converts the collected data into standardized Package URL (PURL) and CPE identifiers.

## Produced Artifacts
The plugin generates a single output file in the build results directory:
- **`<name>-<version>-<release>.sbom`**: A valid CycloneDX 1.5 JSON file containing:
    - Metadata about the build host and distribution.
    - Components for all built packages, including their specific file lists and hashes.
    - Component entries for source code and patches.
    - A dependency graph linking the packages to their runtime requirements.
    - Build environment hardening status (PIE, RELRO, FORTIFY, etc.).

## Configuration
The following options are available in `config.py` to control the plugin behavior:
- **`generate_sbom`** (default: `True`): Enable or disable SBOM generation.
- **`include_file_components`** (default: `True`): Includes individual file entries and their hashes in the SBOM.
- **`include_file_dependencies`** (default: `False`): Maps explicit dependencies between files and their parent packages in the dependency graph.
- **`include_debug_files`** (default: `False`): Determines if debuginfo files are included in the manifests.
- **`include_man_pages`** (default: `True`): Determines if manual pages and documentation are included.
- **`include_source_dependencies`** (default: `True`): Establishes provenance by linking binary packages back to their source files (SRPMs/patches).
- **`include_toolchain_dependencies`** (default: `False`): Includes every package in the build environment as a dependency for full build-root auditing.

## Release Note
Added a new SBOM generator plugin. It creates a `<name>-<version>-<release>.sbom` file in the build results directory, documenting the package components, hashes, signatures, and build environment hardening status.

atomicturtle committed:
Init

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Adding SBOM module

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Update SBOM generator plugin to v1.2.5

This update includes:
- CycloneDX 1.5 schema compliance
- Build hardening metadata capture (FORTIFY, PIE, RELRO, etc.)
- Improved dependency tracking and PURL/CPE support
- Updated documentation for CycloneDX support
- New default configuration options in config.py

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Remove legacy sbom_generator.py from root directory

atomicturtle committed:
Address PR feedback: standardize imports, logging, versioning, and docs

xsuchy commented:
Can you add config option `type` which can have only one option 'cyclonedx`?  But I would love to add later spdx type.

xsuchy commented:
Please add release notes using https://github.com/rpm-software-management/mock/blob/main/docs/Release-Notes-New-Entry.md

atomicturtle committed:
Update license to GPL 2.0 or later

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2024 @@
+# Copyright (C) 2025, Atomicorp, Inc.
+# SPDX-License-Identifier: GPL-2.0-only
```
Yup! No problem

xsuchy commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2025 @@
+        
+        return source_files
+
+    def _convert_to_chroot_path(self, host_path):
```
The name is strange. It does not ring bell to me. This is effectively the reverse function of `buildroot.make_chroot_path()`, right?
I would put it to buildroot.py and call it `from_chroot_path()`.

xsuchy replied:
> self.buildroot.make_chroot_path()

This is **reverse** to what you are doing. self.buildroot.make_chroot_path('FOO') is basically self.buildroot.rootdir + '/' + 'FOO'.

ad ubreq - it does not make this code a function.

The code of your function is correct. I just do not like the name.

xsuchy replied:
Both variants are ok for now. The latter makes a little bit more sense to me.

atomicturtle replied:
Done

praiskup commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2025 @@
+
+    def get_rpm_file_list(self, rpm_path):
+        """Extracts the list of files from an RPM file."""
+        cmd = ["rpm", "-qpl", rpm_path]
```
I don't think this would work reasonably well; we have to account for bootstrapped use-cases.

praiskup commented:
@azhuzhu can you give this a brief look?  How much do we overlap with the syft generator?

atomicturtle committed:
docs: Update SBOM plugin availability to 6.7

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
feat: Refactor SBOM generator and add release notes

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
fix: Use rpm2archive instead of rpm2cpio for source extraction

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
fix: Refactor SBOM generator to use python-rpm and fix local variable error

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
fix: Replace shutil.rmtree with mockbuild.file_util.rmtree for safer cleanup

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
fix: Use python-distro to detect chroot distribution

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
refactor: Rename _convert_to_chroot_path to from_chroot_path

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
refactor: Remove redundant special-case CPE vendor/product mappings

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
fix: Replace deprecated datetime.utcnow() with timezone-aware datetime.now()

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

azhuzhu commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2067 @@
+                                component_obj["licenses"] = [
+                                    {
+                                        "license": {
+                                            "id": license_str
```
It looks `license.id` must be one valid string listing in [CycloneDX v1.5 scheme](https://cyclonedx.org/docs/1.5/json/#components_items_licenses_oneOf_i0_items_license_id). So probably you can use SPDX License Expression instead with `licensfedora2spdx` of [license-validate](https://pagure.io/copr/license-validate)

azhuzhu reviewed (commented):
> @azhuzhu can you give this a brief look?  How much do we overlap with the syft generator? 

Although both are to generate an SBOM for RPM, this plugin is different from the SBOM generator we have.

* This plugin is based on the mock buildroot for one SRPM with a target arch; Our approach is based on an SRPM with multiple arches. So this doesn't have an SRPM Component as the root.
* There are a list of sources and patches here as component items, but in our generator, we only record source archives as Ancestors of the SRPM component
* This plugin also parse runtime deps, which we didn't do.
* This plugin doesn't use other external tools (like `syft`) to scan the content of (S)RPMS and merge the result in the output but we do.
* Some details of the logic are different

atomicturtle committed:
Lint revisions

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Minor refactors

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
SPDX option support

config_opts['sbom_generator_opts'] = {
    'type': 'spdx',
}

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2067 @@
+                    if entry.name.endswith('.rpm') and not entry.name.endswith('.src.rpm'):
+                        rpm_files.append(entry.name)
+                    elif entry.name.endswith('.src.rpm'):
+                        src_rpm_files.append(entry.name)
```
done

atomicturtle committed:
Refactor: Move from_chroot_path to Buildroot class

This moves the path conversion logic from the SBOM generator plugin
to the core Buildroot class, as it is a general-purpose utility
providing the reverse of make_chroot_path().

atomicturtle committed:
modernize metadata extraction and fix forensic grouping

- Refactor sbom_generator to use native python-rpm bindings for metadata extraction
- Fix NameError in _create_built_package_component when accessing sourcerpm
- Improve forensic fidelity by moving signature metadata (date, algorithm) from
  aggregate signer-groups to individual toolchain components
- Resolve merge conflicts with upstream/main in PLUGIN_LIST

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2265 @@
+        # pylint: disable=no-member
+        file_info = {}
+        try:
+            ts = rpm.TransactionSet()
```
I shifted this to the internal model rpmautospec and unbreq are using

atomicturtle committed:
QA fixes:

- Use chroot-native RPM tools via doChroot for internal artifacts
- Fix path mapping to bypass chroot for host-resident result files
- Add regex-based fallback for spec parsing to handle legacy syntax
- Consolidate imports and standardized toolchain PURLs

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Update SBOM docs

- move internal tests out of the project

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Update the Feature request

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Reorganize into separate files

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,2025 @@
+# Copyright (C) 2025, Atomicorp, Inc.
```
Got it, I broke these up into separate files in the plugins/ dir:

mock/py/mockbuild/plugins/sbom_cyclonedx.py 
mock/py/mockbuild/plugins/sbom_spdx.py  
mock/py/mockbuild/plugins/sbom_utils.py

atomicturtle committed:
Bump for build test

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

praiskup commented:
Can you please rebase?

atomicturtle commented:
Anything else you want me to do here?

atomicturtle committed:
Fix CycloneDX SBOM source hashes, metadata, and property regressions

The pre-build spec capture path populated source filenames without digests
and prevented SRPM header extraction from running, leaving most source
artifacts without SHA-256 hashes in the generated SBOM. Always merge SRPM
metadata into spec-derived source entries instead of treating a non-empty
pre-build list as complete.
Restore full CycloneDX document metadata after document creation so build
hardening flags, toolchain tools, and other forensic fields are present in
the final output. Revert source component properties to the expected
mock:source:type "source" and mock:signature:info names used by downstream
auditors.
Add an rpm -qp fallback for SRPM source digest extraction when python-rpm
header parsing fails (for example, when the signing key is unavailable).

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

praiskup commented:
Sorry for the delay on the review.  This is going to get much more attention now, thank you again for your work!

@atomicturtle would you mind crating an executable script doing this business, so we could run it without running the whole mock toolchain?  In mock we would just call the script (we can move the script to a separate project if needed).

praiskup commented on mock/py/mockbuild/plugins/sbom_utils.py:
```diff
@@ -0,0 +1,832 @@
+# -*- coding: utf-8 -*-
```
This file is a library file -- should live in toplevel mockbuild module, to not confuse it with plugin.

atomicturtle replied:
Done

praiskup commented on mock/py/mockbuild/plugins/sbom_utils.py:
```diff
@@ -0,0 +1,832 @@
+                # Verify existence via doChroot
+                cmd = ["ls", candidate]
+                try:
+                    res, _ = self.buildroot.doChroot(cmd, shell=False, returnOutput=True, printOutput=False)
```
The `doChroot` method isn't optimal.  It executes RPM in buildroot, which often isn't working well (even if mock builds fine using RPM from bootstrap or host).

You should be using doOutChroot() (check buildroot_lock that does similar things).

praiskup replied:
Also check package_state.py.

atomicturtle replied:
Done, and done

atomicturtle committed:
Merge upstream/main into sbom_generator

Bring in 14 commits from rpm-software-management/mock main.

atomicturtle committed:
PR feedback changes

1. Moved library components out of plugins dir

	mock/py/mockbuild/plugins/sbom_cyclonedx.py -> mock/py/mockbuild/sbom_cyclonedx.py
	mock/py/mockbuild/plugins/sbom_spdx.py -> mock/py/mockbuild/sbom_spdx.py

2. Created a new CLI generator mock/py/mock-sbom-generator.py, this can
   be run standalone. This does present some forensic gaps vs running
   inline as it is performing operations outside of the build chain.

3. Added a man page

4. Extends forensic coverage to package level sha-256, RSAHEADER, and
   src.rpm GPG checking.

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle commented:
This update changes:

    1. Moved library components out of plugins dir
    
            mock/py/mockbuild/plugins/sbom_cyclonedx.py -> mock/py/mockbuild/sbom_cyclonedx.py
            mock/py/mockbuild/plugins/sbom_spdx.py -> mock/py/mockbuild/sbom_spdx.py
    
    2. Created a new CLI generator mock/py/mock-sbom-generator.py, this can
       be run standalone. This does present some forensic gaps vs running
       inline as it is performing operations outside of the build chain.
    
    3. Added a man page
    
    4. Extends forensic coverage to package level sha-256, RSAHEADER, and
       src.rpm GPG checking.

atomicturtle committed:
hanges based on the code rabbit review

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

praiskup commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,351 @@
+                if key in self.conf:
+                    flag = "--" + key.replace("_", "-")
+                    value = "true" if self.conf[key] else "false"
+                    cmd.extend([flag, value])
```
Can we please convert those into a command-line directly in Mock config?
Something like this?
```
# config_opts['plugin_conf']['sbom_generator_opts'] = {
#     'generate_sbom': True,
#     'type': 'cyclonedx',  # or 'spdx'
#     'command': '/usr/bin/mock-sbom-generator --type cyclonedx --include-file-components etc...',
```
This would seem more flexible (ie., the generator can live in Mock as long as we feel it scales well, but not necessarily).

atomicturtle replied:
Got it, I changed:

- Removed that programmatic flag-building loop
- command is now a full argv template (default in the plugin, documented in site-defaults.cfg), expanded with %() then shlex.split
- (optionally) - Users can replace command with any generator while still getting Mock path/env injection

praiskup commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,351 @@
+LEGACY_PREBUILD_STATE_FILENAME = ".sbom-prebuild.json"
+
+
+@traceLog()
```
no need to use traceLog: #1681

atomicturtle replied:
done

praiskup commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,351 @@
+            if not candidate:
+                continue
+            base = os.path.basename(candidate)
+            if base in ("site-defaults.cfg", "logging.ini"):
```
Are you searching for `config_file`? https://github.com/rpm-software-management/mock/blob/6f7c49f51a732a31084fcf1d589527dd92def3cb/mock/py/mockbuild/config.py#L1064

praiskup replied:
If you attempt to do export a different thing, please fix the docstring.

atomicturtle replied:
_resolve_config_file now prefers config_opts['config_file'] first (the path Mock sets in config.py). The site-defaults.cfg / logging.ini skip of config_paths is only a fallback if that key is missing or the file isn’t there.

praiskup commented on mock/py/mockbuild/plugins/sbom_generator.py:
```diff
@@ -0,0 +1,351 @@
+            isolation = "nspawn" if use_nspawn else "simple"
+
+        env["isolation"] = str(isolation)
+        env["use_nspawn"] = bool(use_nspawn)
```
I am curious if we want to export the config entirely, something like --debug-config or --debug-config-expanded.

atomicturtle replied:
yeah I can see getting that in general, what I was trying to answer here was if mock had rpmbuild_networking enabled or not when the binary was being generated. Im sure there are a bunch of other attack vectors we could isolate by capturing the whole config though

praiskup commented on mock/py/mockbuild/sbom_cyclonedx.py:
```diff
@@ -0,0 +1,979 @@
+# vim:expandtab:autoindent:tabstop=4:shiftwidth=4:filetype=python:textwidth=0:
+# License: GPL2 or later see COPYING
+# Written by Scott R. Shinn <scott@atomicorp.com>
+# Copyright (C) 2026, Atomicorp, Inc.
```
I'd juse SPDX format (like in spec) for licenses, ie GPL-2.0-or-later (but FWIW, I don't think we need these `#` headers at all these days).

atomicturtle replied:
Got it

praiskup commented on mock/py/mock-sbom-generator.py:
```diff
@@ -0,0 +1,344 @@
+        from mockbuild.util import host_path_to_chroot_path
+        return host_path_to_chroot_path(host_path, self.rootdir)
+
+    def doOutChroot(self, command, *args, **kwargs):  # pylint: disable=invalid-name,unused-argument
```
I meant that you could use the Mock's built-in doOutChroot method, not a re-defined one.

atomicturtle replied:
Got it,  plugin now uses Mock’s real Buildroot.doOutChroot for bootstrap-native mode

praiskup replied:
Do we still need this method defined, then?

I'm struggling to understand how the SBOM generation tool/logic should be called:
a) on the host, b) in the bootstrap (if enabled), or c) inside the chroot?

My understanding is that you prefer a), with certain tools being called inside bootstrap chroot (via doOutChroot(), which falls back to on-host).

atomicturtle replied:
Yes, we still need it but only for the standalone CLI:

a) When bootstrap is off (and always for the standalone CLI), the generator runs on the host. Nothing from a Mock chroot is invoked. That makes sense if you don’t trust the chroot, like if networking was allowed and the build pulled in golang deps or similar. Tradeoff: host vs target RPM mismatches, and it can under-report macros / distro quirks (things we score in sbom-auditor, a related but separate project). TL;DR: the SBOM is more trusted, but it has lower forensic fidelity.

b) Bootstrap — Preferred when bootstrap is on: the plugin stages the host-trusted generator into the bootstrap chroot and runs it via Mock’s real Buildroot.doOutChroot (bootstrap-native rpm against the target --root). This is much  more accurate on complex specs that use nested macros and incompatible target RPM DBs. Tradeoff: slightly less trust / more moving parts than pure host. TL;DR: wins on correctness; loses a bit on trust and simplicity. For most Mock builds (bootstrap on, including typical Koji setups), (b) is the right answer; (a) is the fallback / standalone / host-only-trust case.

c) In the target buildroot, we'd never install or run the generator where %build or %check get invoked. 

This duck-typed doOutChroot is only the CLI’s stand-in for (a). On the Mock plugin path with bootstrap enabled, we use real Buildroot.doOutChroot.

praiskup commented on mock/py/mockbuild/sbom_utils.py:
```diff
@@ -0,0 +1,1598 @@
+        Cryptographic verification is handled separately by verify_rpm_signature.
+        """
+        # pylint: disable=no-member,protected-access
+        ts = rpm.TransactionSet()
```
To get this working well, we need the script to be automatically copied inside the bootstrap chroot, and use the "native RPM".

There are basically two important things the plugin should do in my opinion.  Make sure that the right SBOM tool is well-installed in the bootstrap chroot (or on host, if boostrap=off), and execute it (with the arguments user prefers).

praiskup replied:
I mean, there's no precedent for "copy-pasting-script into a (bootstrap) chroot", but we were discussing this several times ... so it might be the time to do it now?

Btw., we always found a way-around it by running `rpm --root` (doOutChroot) like [this](https://github.com/rpm-software-management/mock/blob/6f7c49f51a732a31084fcf1d589527dd92def3cb/mock/py/mockbuild/plugins/package_state.py#L74)

atomicturtle replied:
Got it (this was a pretty big change)

atomicturtle committed:
PR review changes:

- Use a full %()-style command argv template with shlex.split (rpkg_preprocessor pattern)
- drop @traceLog from SBOM modules
- set SPDX GPL-2.0-or-later,
- prefer config_file for provenanced
- clarify StandaloneContext.doOutChroot as a host stand-in for Buildroot.doOutChroot.

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

atomicturtle committed:
Bootstrap native changes

Keep RPM inventory and --checksig on bootstrap-native rpm (never in the
target), while capturing host forensics and hardening macros on the Mock
host and passing them via sbom-prebuild.json so bootstrap execution does
not under-report SELinux, hostname, or build flags.

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>

praiskup commented on mock/py/mockbuild/buildroot.py:
```diff
@@ -195,6 +195,10 @@ def make_chroot_path(self, *paths):
 
+    def from_chroot_path(self, host_path):
+        """Convert an absolute host path into the corresponding path inside the build chroot."""
+        return util.host_path_to_chroot_path(host_path, self.rootdir)
```
Is this change needed?

atomicturtle replied:
Got it, removed

praiskup commented on mock/docs/site-defaults.cfg:
```diff
@@ -302,6 +302,51 @@
+#     'include_debug_files': False,
+#     'include_man_pages': True,
+#     'include_source_dependencies': True,
+#     'include_toolchain_dependencies': False,
```
Imagine there would be other sbom generators available on host in the future.
It would be nice to avoid most of those config options, and modify `command` directly?

atomicturtle replied:
Changed, heres an example:

```
# SPDX instead of CycloneDX
config_opts['plugin_conf']['sbom_generator_opts'] = {
    'generate_sbom': True,
    'command': (
        '/usr/bin/mock-sbom-generator'
        ' --type spdx'
        ' --resultdir %(resultdir)s'
        ' --root %(root)s'
        ' --builddir %(builddir)s'
        ' --include-toolchain-dependencies true'
        ' --online %(online)s'
        ' --rpmbuild-networking %(rpmbuild_networking)s'
        ' --isolation %(isolation)s'
        ' --use-nspawn %(use_nspawn)s'
    ),
}

# Or an external generator later:
# 'command': '/usr/bin/other-sbom-tool --out %(resultdir)s --chroot %(root)s'

```
and this is definitely something going on with the larger E-SBOM mandate, where they have tools analyzing binaries and creating a binary level SBOM analysis. I dont know how I'm going to work those into this yet, but its a good way to leave the design open

atomicturtle committed:
Code review changes

- drop type/include_*/generate_cpe plugin opts so alternate host
generators can be plugged in by editing the command string. Mock
only substitutes path/runtime placeholders.

- Clarify StandaloneContext.doOutChroot as host-only (bootstrap uses
Buildroot.doOutChroot; never the target).

- Remove unused Buildroot.from_chroot_path and test util.host_path_to_chroot_path instead.

- Fix config.py default command so --resultdir/--root are always present.

Signed-off-by: Scott R. Shinn <scott@atomicorp.com>
